File: packages/flutter_sms/android/src/main/java/com/appleeducate/fluttersms/FlutterSmsPlugin.java
package com.appleeducate.fluttersms;

import android.app.Activity;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import io.flutter.plugin.common.PluginRegistry.ActivityResultListener;

/** FlutterSmsPlugin */
public class FlutterSmsPlugin implements MethodCallHandler,  ActivityResultListener {
  private static final int REQUEST_CODE_SEND_SMS = 0315;


  Activity activity;

  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "flutter_sms");
    channel.setMethodCallHandler(new FlutterSmsPlugin(registrar));
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("sendSMS")) {
      if (!canSendSMS()) {
        result.error(
            "device_not_capable",
            "The current device is not capable of sending text messages.",
            "A device may be unable to send messages if it does not support messaging or if it is not currently configured to send messages. This only applies to the ability to send text messages via iMessage, SMS, and MMS.");
        return;
      }

      String message = call.argument("message");
      String recipients = call.argument("recipients");
      sendSMS(result, recipients, message);
      //result.success("SMS Sent!");
    } else if (call.method.equals("canSendSMS")) {
      result.success(canSendSMS());
    } else {
      result.notImplemented();
    }
  }

  private FlutterSmsPlugin(Registrar registrar) {
    this.activity = registrar.activity();
    registrar.addActivityResultListener(this);
  }

  private boolean canSendSMS() {
    if (!activity.getPackageManager().hasSystemFeature(PackageManager.FEATURE_TELEPHONY))
      return false;

    Intent intent = new Intent(Intent.ACTION_SENDTO);
    intent.setData(Uri.parse("smsto:"));
    ActivityInfo activityInfo =
        intent.resolveActivityInfo(activity.getPackageManager(), intent.getFlags());
    if (activityInfo == null || !activityInfo.exported) return false;

    return true;
  }

  private Result result;

  private void sendSMS(Result result, String phones, String message) {
    this.result = result;
    Intent intent = new Intent(Intent.ACTION_SENDTO);
    intent.setData(Uri.parse("smsto:" + phones));
    intent.putExtra("sms_body", message);
    intent.putExtra(Intent.EXTRA_TEXT, message);
    //     intent.putExtra(Intent.EXTRA_STREAM, attachment);
    activity.startActivityForResult(intent, REQUEST_CODE_SEND_SMS);
  }

  @Override
  public boolean onActivityResult(int requestCode, int resultCode, Intent data) {
    if(requestCode == REQUEST_CODE_SEND_SMS && result!=null){
      result.success("finished");
      result = null;
    }
    return true;
  }
}

File: packages/flutter_sms/android/src/main/java/com/appleeducate/fluttersms/FlutterSmsPlugin.kt
package com.appleeducate.fluttersms

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.provider.Settings
import android.util.Log

import java.util.ArrayList

import androidx.core.app.ActivityCompat
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar

/** FlutterSmsPlugin  */
class FlutterSmsPlugin private constructor(internal var activity: Activity) : MethodCallHandler {

    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method == "sendSMS") {
            val message = call.argument<String>("message")
            val recipients = call.argument<ArrayList<String>>("recipients")
            val smsPermission = "SEND_SMS"
            if (!checkPermission(smsPermission)) {
                requestPermission(smsPermission)
                if (!checkPermission(smsPermission)) {
                    openSettings()
                } else {
                    sendSMS(recipients, message)
                }
            } else {
                sendSMS(recipients, message)
            }
            result.success("Sent!")
        } else {
            result.notImplemented()
        }
    }

    private fun sendSMS(phones: ArrayList<String>?, message: String?) {
        val sendIntent = Intent(Intent.ACTION_VIEW)
        sendIntent.putExtra("sms_body", message)
        sendIntent.data = Uri.parse("sms:" + phones!!)
        activity.startActivity(sendIntent)
    }

    private fun checkPermission(permission: String): Boolean {
        var permission = permission
        permission = getManifestPermission(permission)
        Log.i("SimplePermission", "Checking permission : $permission")
        return PackageManager.PERMISSION_GRANTED == ActivityCompat.checkSelfPermission(activity, permission)
    }

    private fun requestPermission(permission: String) {
        var permission = permission
        permission = getManifestPermission(permission)
        Log.i("SimplePermission", "Requesting permission : $permission")
        val perm = arrayOf(permission)
        ActivityCompat.requestPermissions(activity, perm, 0)
    }

    private fun getManifestPermission(permission: String): String {
        val res: String
        when (permission) {
            "SEND_SMS" -> res = Manifest.permission.SEND_SMS
            else -> res = "ERROR"
        }
        return res
    }

    private fun openSettings() {
        val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS,
                Uri.parse("package:" + activity.packageName))
        intent.addCategory(Intent.CATEGORY_DEFAULT)
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
        activity.startActivity(intent)
    }

    companion object {

        /** Plugin registration.  */
        fun registerWith(registrar: Registrar) {
            val channel = MethodChannel(registrar.messenger(), "flutter_sms")
            channel.setMethodCallHandler(FlutterSmsPlugin(registrar.activity()))
        }
    }
}
